src/main/java/com/hierynomus/smbj/transport/tcp/AsyncDirectTcpTransport.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (C)2016 - SMBJ Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hierynomus.smbj.transport.tcp;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.channels.AsynchronousChannelGroup;
+import java.nio.channels.AsynchronousSocketChannel;
+import java.nio.channels.CompletionHandler;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.hierynomus.protocol.Packet;
+import com.hierynomus.protocol.commons.buffer.Buffer;
+import com.hierynomus.protocol.commons.buffer.Buffer.BufferException;
+import com.hierynomus.smbj.lock.CrossThreadLock;
+import com.hierynomus.smbj.transport.PacketHandlers;
+import com.hierynomus.smbj.transport.TransportException;
+import com.hierynomus.smbj.transport.TransportLayer;
+
+/**
+ * A transport layer over Direct TCP/IP that uses asynchronous I/O.
+ */
+public class AsyncDirectTcpTransport<P extends Packet<P, ?>> implements TransportLayer<P> {
+    public static final int DEFAULT_CONNECT_TIMEOUT = 5000;
+
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+    private final PacketHandlers<P> handlers;
+    private final AsynchronousSocketChannel socketChannel;
+    private final AsyncPacketReader<P> packetReader;
+    private int soTimeout = 0;
+
+    // AsynchronousSocketChannel doesn't support concurrent writes
```

Comment:
Do I understand it correctly that this implementation returns early before the write has completed? If so, then the behaviour of the regular synchronous transport and async one are very different and you can't use one as a drop-in replacement for the other.

@hierynomus The documentation of `TransportLayer#write(P packet)` doesn't state it's post-conditions explicitly, which it probably should. Upon return, should the packet have been written out to the network? If not, how do you get notification of this? If that guarantee should be given, then `AsyncDirectTcpTransport#write` might as well be made synchronized and block until the write is complete.